### ISSUE TITLE
ExplicitNonNullaryApply: insert () for non-nullary method application

### ISSUE THREAD

**Issue by olafurpg** [scalacenter/scalafix#204]
Dotty will require explicit parentheses () for non-nullary method applications, see https://github.com/lampepfl/dotty/issues/2571

It would be nice to have a scalafix rewrite ExplicitNullaryApply that produces diffs like this

```diff
- List.newBuilder[Int].result
+ List.newBuilder[Int].result()
```

For a larger example, see https://github.com/lampepfl/dotty/pull/2716

Currently blocked by https://github.com/scalameta/scalameta/issues/930

**Comment by olafurpg** [comment]
Implemented the rewrite here https://github.com/scalacenter/scalafix/compare/master...olafurpg:no-auto-apply?expand=1 Just waiting for release of https://github.com/scalameta/scalameta/pull/1043 to get `Denotation.isJavaDefined`

**Comment by olafurpg** [comment]
I continued working on this issues but got stuck on an issue with Term.ApplyType

https://github.com/scalacenter/scalafix/compare/master...olafurpg:no-auto-apply?expand=1#diff-4b9700111672b3f4b8593bb6dcbbf614R36

We're very close to solving this, but I feel the denotation API is still a bit too raw for these rewrites.

**Comment by olafurpg** [comment]
For methods I'd love to get some sort of Defn.Def structure. Currently I do `Denotation.info.contains("()") to check if a  method is nullary, I first did `.startsWith("())` but that fails with `[T]()`. Implementing rewrites against the strings is too fragile IMO.

**Comment by olafurpg** [comment]
Made a lot of progress on this rewrite today. The logic is sound and I've tested it on scala-library with nice success. There is one major blocker however, the rewrite is triggered for nullary scala 2.x defined methods that override java methods. The scala compiler doesn't store this information in its signatures, so it's not possible to work around this except to fix it in scala/scala. See https://github.com/scalacenter/scalafix/commit/c06015b76108835d41c39f74a793a5c8d4d25cdd

**Comment by olafurpg** [comment]
A possible workaround this blocker issue would be to skip methods that override non-nullary methods. This would avoid false-positives at the cost of introducing a few potential false-negatives.

**Comment by olafurpg** [comment]
> However can you rewrite the call site when the receiver type is a structural type (as shown in the example above)?

From https://github.com/lampepfl/dotty/issues/4026#issuecomment-369229874

**Comment by olafurpg** [comment]
Migration rules for Dotty will live in a different directory than the scalacenter/scalafix so that they can evolve at a different pace than Scalafix APIs. Thanks to recent improvements, custom rules can still have the identical UX as built-in rules.

- automatically installed by `sbt-scalafix`
- tab completions in the sbt shell like built-in rules
- documentation on the main Scalafix website
